test/LeanCode.Test.Helpers/ExternalServiceFactAttribute.cs

```diff
@@ -0,0 +1,39 @@
+using System.Reflection;
+using Xunit;
+using Xunit.v3;
+
+namespace LeanCode.Test.Helpers;
+
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+public abstract class ExternalServiceFactAttribute : FactAttribute, ITraitAttribute, IBeforeAfterTestAttribute
+{
+    protected abstract IReadOnlyCollection<string> RequiredEnvVariables { get; }
+    protected abstract string ServiceType { get; }
+
+    public ExternalServiceFactAttribute()
+    {
+        Explicit = true;
+    }
+
+    public IReadOnlyCollection<KeyValuePair<string, string>> GetTraits() =>
+        [new("category", "external"), new("service", ServiceType), new("integration", "true")];
+
+    public void After(MethodInfo methodUnderTest, IXunitTest test) { }
+
+    public void Before(MethodInfo methodUnderTest, IXunitTest test)
+    {
+        var skipReason = SkipIfVariablesNotSet(RequiredEnvVariables.ToHashSet());
+
+        if (skipReason != null)
+        {
+            Assert.Fail(skipReason);
```

Comment:
Do we really prefer to fail here rather than actually skip?

Reply:
Yes. You now have to explicitly run them (i.e. pass `--explicit on` to the runner), so failing is IMO more sensible. My thinking here is "you already did something to run them, so let's run them".

Reply:
Makes sense.